### PR TITLE
Course work

### PR DESCRIPTION
**Pull request by Oleh325**

### CODE REVIEW

File: src/main/java/ua/lviv/iot/novaPoshtaAPI/controller/CourierController.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import ua.lviv.iot.novaPoshtaAPI.service.CourierService;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/couriers")
+public class CourierController {
+
+    @Autowired
+    CourierService courierService;
```
Comment (zveres-ss) [inline src/main/java/ua/lviv/iot/novaPoshtaAPI/controller/CourierController.java]: missed visibility modifier

File: src/main/java/ua/lviv/iot/novaPoshtaAPI/controller/CourierController.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.DeleteMapping;
+import ua.lviv.iot.novaPoshtaAPI.model.Courier;
+import ua.lviv.iot.novaPoshtaAPI.service.CourierService;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/couriers")
+public class CourierController {
```
Comment (zveres-ss) [inline src/main/java/ua/lviv/iot/novaPoshtaAPI/controller/CourierController.java:22]: don't see the API to create Couriers or update their data

File: src/main/java/ua/lviv/iot/novaPoshtaAPI/controller/CourierController.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    public List<Courier> getAllCouriers() {
+        return courierService.getAllCouriers();
+    }
+
+    @GetMapping("/{courierId}")
+    public Courier getCourierById(@PathVariable Long courierId) {
+        return courierService.getCourierById(courierId);
+    }
+
+    @RequestMapping("/{courierId}/deliver/{parcelId}")
```
Comment (zveres-ss) [inline src/main/java/ua/lviv/iot/novaPoshtaAPI/controller/CourierController.java]: this method is operating with parcel, not courier. So it should be in ParcelController (together with methods to create the parcel as well)

File: src/main/java/ua/lviv/iot/novaPoshtaAPI/controller/DepartmentController.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+import ua.lviv.iot.novaPoshtaAPI.service.DepartmentService;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/departments")
+public class DepartmentController {
+
+    @Autowired
+    DepartmentService departmentService;
```
Comment (zveres-ss) [inline src/main/java/ua/lviv/iot/novaPoshtaAPI/controller/DepartmentController.java]: missed visibility modifier

File: src/main/java/ua/lviv/iot/novaPoshtaAPI/controller/DepartmentController.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+    @GetMapping
+    private List<Department> getAllDepartments() {
```
Comment (zveres-ss) [inline src/main/java/ua/lviv/iot/novaPoshtaAPI/controller/DepartmentController.java]: this method is private, nobody will be able to call it

File: src/main/java/ua/lviv/iot/novaPoshtaAPI/controller/DepartmentController.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        departmentService.updateDepartment(department, departmentId);
+    }
+
+    @DeleteMapping("/{departmentId}")
+    public void deleteDepartment(@PathVariable Long departmentId) {
+        departmentService.deleteDepartment(departmentId);
+    }
+
+    @RequestMapping("/{departmentId}/couriers/{courierId}/parcels/{parcelId}")
+    public void giveParcelToCourier(@PathVariable Long departmentId, @PathVariable Long courierId,
```
Comment (zveres-ss) [inline src/main/java/ua/lviv/iot/novaPoshtaAPI/controller/DepartmentController.java]: this code belongs to ParcelController

File: src/main/java/ua/lviv/iot/novaPoshtaAPI/controller/DepartmentController.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+                                    @PathVariable Long parcelId) {
+        departmentService.giveParcelToCourier(departmentId, courierId, parcelId);
+    }
+
+    @GetMapping("/{departmentId}/parcels")
```
Comment (zveres-ss) [inline src/main/java/ua/lviv/iot/novaPoshtaAPI/controller/DepartmentController.java]: move the methods related to parcels to parcel controller

File: src/main/java/ua/lviv/iot/novaPoshtaAPI/controller/DepartmentController.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    @RequestMapping("/{departmentIdFrom}/deliver/{departmentIdTo}/parcels/{parcelId}")
+    public void deliverParcel(@PathVariable Long departmentIdFrom, @PathVariable Long departmentIdTo,
+                              @PathVariable Long parcelId) {
+        departmentService.deliverParcel(departmentIdFrom, departmentIdTo, parcelId);
+    }
+
+    @PostMapping("/couriers")
+    public void addCourier(@RequestBody Courier courier) {
```
Comment (zveres-ss) [inline src/main/java/ua/lviv/iot/novaPoshtaAPI/controller/DepartmentController.java]: move method to courierController

File: src/main/java/ua/lviv/iot/novaPoshtaAPI/datastorage/dal/CourierFileStore.java [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+        List<Courier> resultList = new LinkedList<>();
+
+        File directory = new File("res");
+        if (!directory.exists()) {
+            directory.mkdir();
+        }
+
+        String testPath = "";
+        if (isTest) {
+            testPath = "test\\";
```
Comment (zveres-ss) [inline src/main/java/ua/lviv/iot/novaPoshtaAPI/datastorage/dal/CourierFileStore.java]: production code should not use any variables for tests or test related logic

File: src/main/java/ua/lviv/iot/novaPoshtaAPI/datastorage/dal/CourierFileStore.java [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+                    file = new File("res\\" + testPath + "courier-" + year + "-"
+                            + month + "-0" + i + ".csv");
+                    resultList.addAll(scanCourier(file));
+                }
+            } else {
+                if (Files.exists(Paths.get("res\\" + testPath + "courier-" + year + "-"
+                        + month + "-" + i + ".csv"))) {
+                    file = new File("res\\" + testPath + "courier-" + year + "-"
+                            + month + "-" + i + ".csv");
+                    resultList.addAll(scanCourier(file));
```
Comment (zveres-ss) [inline src/main/java/ua/lviv/iot/novaPoshtaAPI/datastorage/dal/CourierFileStore.java]: please use somehow src/main/java/ua/lviv/iot/novaPoshtaAPI/Util.java logic here

File: src/main/java/ua/lviv/iot/novaPoshtaAPI/datastorage/dal/CourierFileStore.java [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+                }
+            }
+        }
+
+        return resultList;
+    }
+
+    private List<Courier> scanCourier(File file) throws IOException {
```
Comment (zveres-ss) [inline src/main/java/ua/lviv/iot/novaPoshtaAPI/datastorage/dal/CourierFileStore.java]: please refactor this method - its size is bigger than one screen

File: src/main/java/ua/lviv/iot/novaPoshtaAPI/datastorage/dal/CourierFileStore.java [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.time.LocalDate;
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Component
+public class CourierFileStore {
```
Comment (zveres-ss) [inline src/main/java/ua/lviv/iot/novaPoshtaAPI/datastorage/dal/CourierFileStore.java]: ok, can't see where is a HashMap to cache the data from file, not to read it from file each time

File: src/main/java/ua/lviv/iot/novaPoshtaAPI/datastorage/dal/DepartmentFileStore.java [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+            } else {
+                if (Files.exists(Paths.get("res\\" + testPath + "department-" + year + "-"
+                        + month + "-" + i + ".csv"))) {
+                    file = new File("res\\" + testPath + "department-" + year + "-"
+                            + month + "-" + i + ".csv");
+                    resultList.addAll(scanDepartment(file));
+                }
+            }
+        }
+
```
Comment (zveres-ss) [inline src/main/java/ua/lviv/iot/novaPoshtaAPI/datastorage/dal/DepartmentFileStore.java]: logic is duplicated in src/main/java/ua/lviv/iot/novaPoshtaAPI/datastorage/dal/CourierFileStore.java

the file name logic should be moved to either parent class or Util

File: src/main/java/ua/lviv/iot/novaPoshtaAPI/datastorage/dal/ParcelFileStore.java [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+
+        File file = new File("res\\" + testPath + "parcel-" + date + ".csv");
+        try (Writer writer = new OutputStreamWriter(new FileOutputStream(file), StandardCharsets.UTF_8);) {
+            writer.write(parcels.get(0).getHeaders() + "\n");
+            for (Parcel parcel : parcels) {
+                writer.write(parcel.toCSV() + "\n");
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
```
Comment (zveres-ss) [inline src/main/java/ua/lviv/iot/novaPoshtaAPI/datastorage/dal/ParcelFileStore.java]: do not suppress exceptions - your code will not work if it occurs

File: src/main/java/ua/lviv/iot/novaPoshtaAPI/service/DepartmentService.java [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+
+@Service
+public class DepartmentService {
+
+    @Autowired
+    ParcelService parcelService;
```
Comment (zveres-ss) [inline src/main/java/ua/lviv/iot/novaPoshtaAPI/service/DepartmentService.java]: use visibility modifiers

File: src/main/java/ua/lviv/iot/novaPoshtaAPI/service/DepartmentService.java [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+            newDpt.setLocation(department.getLocation());
+        } else {
+            newDpt.setLocation(oldDpt.getLocation());
+        }
+        if (department.getWorkingHours() != null) {
+            newDpt.setWorkingHours(department.getWorkingHours());
+        } else {
+            newDpt.setWorkingHours(oldDpt.getWorkingHours());
+        }
+
```
Comment (zveres-ss) [inline src/main/java/ua/lviv/iot/novaPoshtaAPI/service/DepartmentService.java]: imagine your application crashed at this point. You dropped the old department while a new one is not stored. hash map has an ability to replace elements with the same key automatically

File: src/main/java/ua/lviv/iot/novaPoshtaAPI/service/ParcelService.java [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+
+    public Parcel getParcelById(Long parcelId) {
+        return this.parcels.get(parcelId);
+    }
+
+    public void addParcel(Parcel parcel) {
+        parcels.put(parcel.getParcelId(), parcel);
+    }
+
+    public void updateParcel(Parcel parcel, Long parcelId) {
```
Comment (zveres-ss) [inline src/main/java/ua/lviv/iot/novaPoshtaAPI/service/ParcelService.java:36]: you just replacing the old parcwl with new one. Validation could be done by annotations on the model level. Your logic is unclear to the API caller - for some reasons passed fields are not overrriden but simply taken from old object

File: src/main/java/ua/lviv/iot/novaPoshtaAPI/Util.java [hunk 18]
```diff
@@ -0,0 +1,25 @@
+package ua.lviv.iot.novaPoshtaAPI;
+
+import java.time.LocalDate;
+
+public class Util {
```
Comment (zveres-ss) [inline src/main/java/ua/lviv/iot/novaPoshtaAPI/Util.java]: this class is not used in the code

**Review by zveres-ss — CHANGES_REQUESTED** [review]

File: src/main/java/ua/lviv/iot/novaPoshtaAPI/controller/CourierController.java (hunk 2 above)
Comment (Oleh325) [reply src/main/java/ua/lviv/iot/novaPoshtaAPI/controller/CourierController.java:22]: Couriers can only be created and updated via departments, so there won't be any "hanging" couriers without a department

File: src/main/java/ua/lviv/iot/novaPoshtaAPI/controller/CourierController.java (hunk 3 above)
Comment (Oleh325) [reply src/main/java/ua/lviv/iot/novaPoshtaAPI/controller/CourierController.java]: See the reply above. The logic here has been changed a bit from a traditional one

File: src/main/java/ua/lviv/iot/novaPoshtaAPI/datastorage/dal/CourierFileStore.java (hunk 10 above)
Comment (Oleh325) [reply src/main/java/ua/lviv/iot/novaPoshtaAPI/datastorage/dal/CourierFileStore.java]: it is used in saving logic. Loading one is different(we don't use "-day" here, from Util getDateNow)

File: src/main/java/ua/lviv/iot/novaPoshtaAPI/Util.java (hunk 18 above)
Comment (Oleh325) [reply src/main/java/ua/lviv/iot/novaPoshtaAPI/Util.java]: it is used, but I'll add one more method with next commit

File: src/main/java/ua/lviv/iot/novaPoshtaAPI/Util.java [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+        if (!directory.exists()) {
+            directory.mkdirs();
+        }
+    }
+
+    public static List<String> processRawValues(List<String> rawValues) {
+        List<String> values = new LinkedList<>();
+        boolean isList = false;
+        String listValue = "";
+        for (String value : rawValues) {
```
Comment (zveres-ss) [inline src/main/java/ua/lviv/iot/novaPoshtaAPI/Util.java]: overcomplicated logic, please store one field per one column

**Review by zveres-ss — CHANGES_REQUESTED** [review]

**Pull request merged**